packages/loader/container-loader/src/virtualDeltaManager.ts

```diff
@@ -0,0 +1,278 @@
+                clientSequenceNumber: this.connectionManager.clientId === message.clientId ? this.virtualizeClientSequenceNumber(message.clientSequenceNumber) : message.clientSequenceNumber,
+                referenceSequenceNumber: this.virtualizeSequenceNumber(message.referenceSequenceNumber),
+                minimumSequenceNumber: this.virtualizeSequenceNumber(message.minimumSequenceNumber),
+            }, startTime);
+        }
+    }
+
+    private virtualizeSequenceNumber(sequenceNumber: number): number {
+        if (sequenceNumber < 0) {
+            return sequenceNumber;
+        }
+        const virtualizedSequenceNumber = this._sequenceNumberMap.get(sequenceNumber);
+        assert(virtualizedSequenceNumber !== undefined, "sequenceNumber not found");
+
+        return virtualizedSequenceNumber;
+    }
+
+    private virtualizeClientSequenceNumber(clientSequenceNumber: number): number {
+        if (clientSequenceNumber < 0) {
+            return clientSequenceNumber;
+        }
+        const virtualizedClientSequenceNumber = this._clientSequenceNumberMap.get(clientSequenceNumber);
+        assert(virtualizedClientSequenceNumber !== undefined, "clientSequenceNumber not found");
+
+        return virtualizedClientSequenceNumber;
+    }
+
+    private virtualizeSequencedMessage(message: ISequencedDocumentMessage, storeValue: boolean = false): ISequencedDocumentMessage {
+        if (storeValue) {
+            if (message.type === GroupedBatchOpType) {
+                this._virtualSequenceNumber += (message.contents as IDocumentMessage[]).length;
+                this._sequenceNumberMap.set(message.sequenceNumber, this._virtualSequenceNumber); // We store last sequenceNumber of group
+            } else {
+                this._sequenceNumberMap.set(message.sequenceNumber, ++this._virtualSequenceNumber);
+            }
+        }
+        return {
+            ...message,
+            sequenceNumber: this.virtualizeSequenceNumber(message.sequenceNumber),
+            referenceSequenceNumber: this.virtualizeSequenceNumber(message.referenceSequenceNumber),
+            minimumSequenceNumber: this.virtualizeSequenceNumber(message.minimumSequenceNumber),
+            clientSequenceNumber: this.connectionManager.clientId === message.clientId ? this.virtualizeClientSequenceNumber(message.clientSequenceNumber) : message.clientSequenceNumber,
+        };
+    }
+
+    // TODO: need to review how these numbers are being used in other layers
+    // there is currently some mix and match happening
+
+    // TODO: virtualize this number
+    public get initialSequenceNumber(): number {
+        return this.initSequenceNumber;
+    }
+
+    public get lastSequenceNumber(): number {
+        return this.virtualizeSequenceNumber(this.lastProcessedSequenceNumber);
+    }
+
+    public get lastMessage() {
+        return this.lastProcessedMessage !== undefined ? this.virtualizeSequencedMessage(this.lastProcessedMessage) : undefined;
```

Comment:
Isn't it double-virtualizaiton? I.e. this.lastProcessedMessage content is result of inbound queue processing.

Reply:
Not technically. The _inbound queue stores the un-virtualized values. It's only when other places try to ask for those values where they get virtualized. All the values are un-virtualized in DeltaManager